### PR TITLE
fix(`evm`): check for global failure instead of  DSTest flag when reverting snapshots

### LINKED ISSUE #3792: Test reverts with `Assertion failed.` when reverting to a fork snapshot, without any assertion
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (15204ec 2022-11-29T00:07:45.954046141Z)

### What command(s) is the bug in?

forge test

### Operating System

Linux

### Describe the bug

# Report

- The following test case `testWtf` reverts with reason `Assertion failed.`, whereas no assertion is explicitly performed within it

```Solidity
// SPDX-License-Identifier: GNU AGPLv3
pragma solidity ^0.8.0;

import "@forge-std/Test.sol";

contract Config {
    address public test = 0xcBa28b38103307Ec8dA98377ffF9816C164f9AFa;
}

contract TestSetup is Config, Test {
    function testWtf() public {
        uint256 snapshotId = vm.snapshot();

        vm.prank(test);

        vm.revertTo(snapshotId);
    }
}
```

- `forge-std` is at the latest commit: `ea1c1f4022f6a7924887106caa64499317295493`

# Findings

It seems to be a very specific case, because the test passes as soon as:
- the order of inheritance is swapped
- the `test` address is moved to the `TestSetup` contract
- the `test` address is `address(1)`, `address(2)`, ...
- any of the line of the body of `testWtf` is commented out

The test fails even if a call is performed on behalf of the pranked address `test`, right after `vm.prank`

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

Closes #3792.

The root issue of these soundness issues comes from storage layout shifts depending on the inheritance order of the test contract. We, right now, check for the `_failed` flag on `DSTest`, which, assuming the user has the correct inheritance order, is a valid way to ensure we're keeping track of snapshot failures. However, if it's not correct, the flag will not be at the expected slot, and therefore we'll get unsound behavior.

## Solution

What we do instead is now rely on the global failure slot that dstest keeps on the hevm address (`keccak256("failed")`) to check for failures. This should avoid any unsoundness issues regarding the inheritance order when it comes to snapshots.

### CODE REVIEW

File: evm/src/executor/backend/mod.rs [hunk 1]
```diff
@@ -873,12 +873,15 @@ impl DatabaseExt for Backend {
...
+            // We first check if there's a "global failure", meaning that the ds-test contract
+            // stored a failure on hevm's keccak256("failed") slot.
+            // We do this instead of checking the usual _failed slot to avoid false positives or
+            // negatives due to any shift in storage due to a wrong inheritance order.
             if self
                 .test_contract_address()
                 .map(|addr| self.is_failed_test_contract_state(addr, current_state))
-                .unwrap_or_default()
+                .unwrap_or_default() ||
+                self.is_global_failure()
```
Comment (Evalir) [inline evm/src/executor/backend/mod.rs:884]: Hey @ZeroEkkusu thanks for the pointer here on #3792 king—I was wondering if this looks correct to you? We're essentially extending the check to look for global failures too.

**Review by Evalir — COMMENTED** [review]
I'm a bit suspicious of the `is_global_failure` function. While debugging, the "failed" slot should absolutely be set to 1 on repro `3055`, but the function is not returning this. However, logging the slot with dstest will show that the slot is indeed 1. Will check if there's something wrong in this func tomorrow

File: evm/src/executor/backend/mod.rs [hunk 2]
```diff
@@ -65,6 +65,9 @@ type ForkLookupIndex = usize;
 const DEFAULT_PERSISTENT_ACCOUNTS: [H160; 3] =
     [CHEATCODE_ADDRESS, DEFAULT_CREATE2_DEPLOYER, CALLER];
 
+const GLOBAL_FAILURE_SLOT: &str =
```
Comment (mattsse) [inline evm/src/executor/backend/mod.rs:68]: needs docs that this is keccak("failed")

**Review by mattsse — CHANGES_REQUESTED** [review]
nice find!

needs some docs for the slot

File: evm/src/executor/backend/mod.rs (hunk 2 above)
Comment (Evalir) [reply evm/src/executor/backend/mod.rs:69]: done! https://github.com/foundry-rs/foundry/pull/5404/commits/a75f9fbf5d790b737240ba89bcdc15b30307d157

a note: this is actually only "failed" in bytes and padded—not `keccak("failed")`

**Review by mattsse — APPROVED** [review]

**Pull request merged**